lib/matplotlib/cbook.py

```diff
@@ -2358,6 +2382,12 @@ def _unpack_to_numpy(x):
         # so in this case we do not want to return a function
         if isinstance(xtmp, np.ndarray):
             return xtmp
+    if _is_torch_array(x) or _is_jax_array(x):
+        xtmp = x.__array__()
```

Comment:
Is there any reason not to do `xtmp = np.asarray(x)` here?

Reply:
I think these are equivalent for the handled cases. While [pandas claims](https://pandas.pydata.org/docs/reference/api/pandas.Series.__array__.html#pandas.Series.__array__) you shouldn't call `__array__` directly, I haven't found any official recommendation for it in numpy (which defines the `__array__` API).

For me, either works. `__array__` is more explicit, which can be a good thing, but might be too low level. OTOH I don't think `np.asarray()` will change the implementation so that the `asarray` abstraction would be safer.

Reply:
@jakevdp @timhoffm What's your opinion about using `x.numpy(force=True)` or `x.detach().cpu().numpy()` for PyTorch? As a user, I'd find this change useful for day-to-day coding since it saves me from manually writing it for every array I want to plot with matplotlib.

Reply:
Seems reasonable to use the `x.numpy` method provided by torch. Though I have to say, I don't know whether `__array__` would do something different.

Reply:
@timhoffm To add more context, in JAX, `.__array__()` method converts a JAX array to a NumPy array irrespective of whether the JAX array is on CPU, GPU, TPU and probably other hardware accelerators. OTOH in PyTorch, they don't do it by default due to uncertainty about performance impact (full discussion is in [this issue](https://github.com/pytorch/pytorch/issues/20778)). So, when we do `x.numpy(force=True)`, it forcefully converts it to NumPy irrespective of the device of the array (handles few others cases as well, e.g. if array has a computation graph for backpropagation then `x.detach()` needs to be called first). I am not sure if this should be handled in a separate issue or we can use `x.numpy(force=True)` in this PR itself but I'm sure that PyTorch users would love this change to avoid writing `x.detach().cpu().numpy()` every time they plot a PyTorch array.

cc @jakevdp.

Reply:
My hesitation about this is that memory overflows will get shunted to Matpolotlib, and then we will get the bug reports, whereas if people did `x.*.numpy()` in their own code, they would see what the problem is.  Jax arrays can be far larger than memory allows, and Matplotlib blindly unpacking them for naive users seems like a bad idea.

Reply:
> Though I have to say, I don't know whether `__array__` would do something different.

`__array__ is the last thing tried by NumPy when a user would call `np.asarray`, after the buffer protocol and `__array_interface__` (as documented for example [here](https://numpy.org/devdocs/user/basics.interoperability.html#using-arbitrary-objects-in-numpy)). So they're not equivalent, and calling `np.asarray` is idiomatic.